File: frontend/src/lib/api/progress.ts
import apiClient from './axios'

export interface LessonProgress {
  id: string;
  user_id: string;
  course_id: string;
  lesson_id: string;
  is_completed: boolean;
  completed_at?: string;
  last_watched_at: string;
  watch_duration_seconds: number;
  created_at: string;
  updated_at: string;
}

export interface CourseProgress {
  course_id: string;
  total_lessons: number;
  completed_lessons: number;
  progress_percent: number;
  last_lesson_id?: string;
  lesson_progress: LessonProgress[];
}

export interface ProgressResponse {
  success: boolean;
  message: string;
  data: CourseProgress;
}

export interface MarkCompletedResponse {
  success: boolean;
  message: string;
  data: {
    lesson_id: string;
  };
}

export interface UpdateWatchProgressResponse {
  success: boolean;
  message: string;
  data: {
    lesson_id: string;
    duration_seconds: number;
  };
}

// Get course progress
export async function getCourseProgress(courseId: string): Promise<ProgressResponse> {
  const response = await apiClient.get<ProgressResponse>(`/progress/${courseId}`);
  return response.data;
}

// Mark a lesson as completed
export async function markLessonCompleted(courseId: string, lessonId: string): Promise<MarkCompletedResponse> {
  const response = await apiClient.post<MarkCompletedResponse>(
    `/progress/${courseId}/lessons/${lessonId}/complete`
  );
  return response.data;
}

// Update watch progress for a lesson
export async function updateWatchProgress(
  courseId: string,
  lessonId: string,
  durationSeconds: number
): Promise<UpdateWatchProgressResponse> {
  const response = await apiClient.post<UpdateWatchProgressResponse>(
    `/progress/${courseId}/lessons/${lessonId}/watch`,
    { duration_seconds: durationSeconds }
  );
  return response.data;
}

// Update last lesson (when user switches to a new lesson)
export async function updateLastLesson(courseId: string, lessonId: string): Promise<void> {
  await apiClient.post(`/progress/${courseId}/last-lesson/${lessonId}`);
}

File: backend/internal/domain/progress.go
package domain

import (
	"time"

	"github.com/google/uuid"
)

// LessonProgress represents a user's progress on a specific lesson
type LessonProgress struct {
	ID                   uuid.UUID  `json:"id"`
	UserID               uuid.UUID  `json:"user_id"`
	CourseID             uuid.UUID  `json:"course_id"`
	LessonID             uuid.UUID  `json:"lesson_id"`
	IsCompleted          bool       `json:"is_completed"`
	CompletedAt          *time.Time `json:"completed_at,omitempty"`
	LastWatchedAt        time.Time  `json:"last_watched_at"`
	WatchDurationSeconds int        `json:"watch_duration_seconds"`
	CreatedAt            time.Time  `json:"created_at"`
	UpdatedAt            time.Time  `json:"updated_at"`
}

// CourseProgress represents the overall progress of a user in a course
type CourseProgress struct {
	CourseID         uuid.UUID         `json:"course_id"`
	TotalLessons     int               `json:"total_lessons"`
	CompletedLessons int               `json:"completed_lessons"`
	ProgressPercent  int               `json:"progress_percent"`
	LastLessonID     *uuid.UUID        `json:"last_lesson_id,omitempty"`
	LessonProgress   []*LessonProgress `json:"lesson_progress,omitempty"`
}
